### DISCUSSION TITLE
Types guaranteed by where clause asking for not null

### DISCUSSION THREAD

**Discussion by JaKXz** [prisma/prisma#20314] [Q&A]
Hi, I did a quick search couldn't find something related, sorry if this is duplicated though.

I am doing a `findUniqueOrThrow` and providing `not: null` conditions in my where clause and I was hoping that the type of the object that comes back would match the where `not: null` conditions I have set e.g.

```ts
db.table.findUniqueOrThrow({
  where: {
    id, //function param
    columnA: { not: null },
    relationshipBId: { not: null },
  },
  include: { relationshipB: true }
});
```

Right now the only workaround I have is to cast the resultant object into what I know it is but that is not ideal. Has anyone figured out a way to solve this? Thanks!

**Comment by nurul3101** [comment]
Hello @JaKXz 👋 

What type are you currently getting?
Can you please share them?

Based on my understanding, the type of the object that is returned from a `findUnique` or `findUniqueOrThrow` call is determined by the type of the table and not the where clause. This is because the where clause is used to filter which rows are returned, but it wouldn't change the structure or type of those rows.

**Reply by nurul3101** [reply]
@JaKXz Checking to see if you had a chance to look at my last response.

Please let us know in case you still need assistance. If not, please consider marking the discussion as closed/answered.

**Reply by JaKXz** [reply]
Essentially, I get `Prisma.GetResult`s where the columns that I have explicitly specified as `not: null` are still nullable. I think it'd be harder to resolve the `include`d relationship as not null but in this case I have asked for the result where the foreign key for that relationship is not null.

e.g. given the `findUniqueOrThrow` call above, I get something that looks like:

```
GetResult<{ columnA: string | null; relationshipBId: string | null; }> & GetResult<{ relationshipB: (table definition) | null }>
```

Sorry again @nurul3101 for the delay, appreciate you looking at this.

What I would expect is 

```
GetResult<{ columnA: string; relationshipBId: string; }> & GetResult<{ relationshipB: (table definition) }>
```

so that I don't have to manually cast this as an intersection type

**Reply by JaKXz** [reply]
> This is because the where clause is used to filter which rows are returned, but it wouldn't change the structure or type of those rows.

Agreed on this generally, but, in the case of using `not: null` as part of the `where` predicates I think it's reasonable to assert that the value for that column will be a `NonNullable` right?

**Reply by nurul3101** [reply]
Do you mind testing this again on Prisma version `5.1.1.`?
The return types would have slightly changed, the `GetResult` wrapper should go away.

> where predicates I think it's reasonable to assert that the value for that column will be a NonNullable right?
I believe a column being `nullable` or `NonNullable` would depend on whether the column is marked as optional or not in the schema file.

**Reply by JaKXz** [reply]
@nurul3101 maybe my question isn't clear, sorry.

> I believe a column being nullable or NonNullable would depend on whether the column is marked as optional or not in the schema file.

Again, totally agree in the very basic `SELECT TOP 1 * FROM table` case. 

However, when I specify

```
SELECT <columns>
FROM <table>
WHERE columnA IS NOT NULL
AND relationshipBId IS NOT NULL
```

I would the returned object's types to recognize that those columns will not be nullable (since it will not be returned if they are null).

**Reply by sambartik** [reply]
> Based on my understanding, the type of the object that is returned from a findUnique or findUniqueOrThrow call is determined by the type of the table and not the where clause.

As I understand it based on the information in the [following docs page](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/type-utilities?query=type%20utilit&page=1):

> Result<Type, Arguments, Operation>: Takes the input arguments and provides the result for a given model and operation. You would usually use this in conjunction with Args. As with Args, Result helps you to re-use existing types to extend or modify them.


The type of the object returned is defined by the model (table), the operation performed on the model (findUniqueOrThrow in the case of this discussion) and the args supplied to the operation.

**Reply by JaKXz** [reply]
Hey @nurul3101 just wondering if you need any more clarifying information or if anyone on the team is able to help me out? I've since upgraded to v5.1.1 and not seeing any difference in the types.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
